crates/matrix-sdk/src/test_utils/mocks/encryption.rs

```diff
@@ -147,6 +137,48 @@ impl MatrixMockServer {
         (alice, bob)
     }
 
+    /// Creates a third client for e2e tests.
+    pub async fn set_up_carl_for_encryption(&self, alice: &Client, bob: &Client) -> Client {
+        let carl_user_id = owned_user_id!("@carlg:example.org");
+        let carl_device_id = owned_device_id!("CARL_DEVICE");
+
+        let carl =
+            self.client_builder_for_crypto_end_to_end(&carl_user_id, &carl_device_id).build().await;
+
+        // Let carl upload it's device keys.
+        self.mock_sync().ok_and_run(&carl, |_| {}).await;
+
+        // Have Alice track Carl, so she queries his keys later.
+        MatrixMockServer::update_tracked_users(alice, [carl.user_id().unwrap()]).await;
+
+        // Have Bob track Carl, so she queries his keys later.
+        MatrixMockServer::update_tracked_users(bob, [carl.user_id().unwrap()]).await;
+
+        // Have Alice and Bob upload their signed device keys, and download Carl's keys.
+        {
+            self.mock_sync().ok_and_run(alice, |_| {}).await;
+            self.mock_sync().ok_and_run(bob, |_| {}).await;
+        }
+
+        // Let carl be aware of Alice and Bob keys.
+        MatrixMockServer::update_tracked_users(
+            &carl,
+            [alice.user_id().unwrap(), bob.user_id().unwrap()],
+        )
+        .await;
+
+        // A last sync for carl to get the keys.
+        self.mock_sync().ok_and_run(alice, |_| {}).await;
+
+        carl
+    }
+
+    async fn update_tracked_users(client: &Client, user_ids: impl IntoIterator<Item = &UserId>) {
+        let olm = client.olm_machine_for_testing().await;
+        let olm = olm.as_ref().unwrap();
+        olm.update_tracked_users(user_ids).await.unwrap();
+    }
```

Comment:
This could be a function on the `Client` that is guarded with `#[cfg(test)]`, no?

Reply:
right, I knew it was not the best place but I didn't thought about the cfg(test) thing https://github.com/matrix-org/matrix-rust-sdk/pull/5252/commits/912c7ef16947ad8fc30afed66e621574aeef917a